src/kernels/MIOpenInterpolate.cpp

```diff
@@ -0,0 +1,967 @@
+        input, output, input_tv, output_tv, nelems, scale_factors, align_corners);
+}
+
+template <typename TI, typename TD>
+__device__ inline void interpolateBicubicBackward(TD* __restrict__ workspace,
+                                                  const TI* __restrict__ output_grad,
+                                                  const tensor_view_t<4> input_grad_tv,
+                                                  const tensor_view_t<4> output_grad_tv,
+                                                  const size_t nelems,
+                                                  const float* scale_factors,
+                                                  const bool align_corners)
+{
+    int64_t gid = blockIdx.x * blockDim.x + threadIdx.x;
+    if(gid >= nelems)
+        return;
+
+    auto tensor_layout = tensor_layout_t<4>(output_grad_tv, gid);
+    int64_t n          = tensor_layout.layout[0];
+    int64_t c          = tensor_layout.layout[1];
+    int64_t h          = tensor_layout.layout[2];
+    int64_t w          = tensor_layout.layout[3];
+
+    int64_t Hin  = input_grad_tv.size[2];
+    int64_t Hout = output_grad_tv.size[2];
+    int64_t Win  = input_grad_tv.size[3];
+    int64_t Wout = output_grad_tv.size[3];
+
+    if(Hin == Hout && Win == Wout)
+    {
+        workspace[input_grad_tv.get_tensor_view_idx(tensor_layout)] =
+            CVT_FLOAT2ACCUM(output_grad[output_grad_tv.get_tensor_view_idx(tensor_layout)]);
+        return;
+    }
+
+    FLOAT_ACCUM scale_factor_h = CVT_FP32_2ACCUM(scale_factors[0]);
+    FLOAT_ACCUM scale_factor_h_ =
+        compute_linear_scale_factor(scale_factor_h, Hin, Hout, align_corners);
+    FLOAT_ACCUM real_y = bicubic_idx(h, Hout, scale_factor_h_, align_corners);
+    int64_t in_y       = static_cast<int64_t>(floor(real_y));
+    FLOAT_ACCUM t_y    = real_y - static_cast<FLOAT_ACCUM>(in_y);
+
+    FLOAT_ACCUM scale_factor_w = CVT_FP32_2ACCUM(scale_factors[1]);
+    FLOAT_ACCUM scale_factor_w_ =
+        compute_linear_scale_factor(scale_factor_w, Win, Wout, align_corners);
+    FLOAT_ACCUM real_x = bicubic_idx(w, Wout, scale_factor_w_, align_corners);
+    int64_t in_x       = static_cast<int64_t>(floor(real_x));
+    FLOAT_ACCUM t_x    = real_x - static_cast<FLOAT_ACCUM>(in_x);
+
+    FLOAT_ACCUM y_coeffs[4];
+    FLOAT_ACCUM x_coeffs[4];
+    get_cubic_upsampling_coefficients(y_coeffs, t_y);
+    get_cubic_upsampling_coefficients(x_coeffs, t_x);
+
+    FLOAT_ACCUM out_value =
+        CVT_FLOAT2ACCUM(output_grad[output_grad_tv.get_tensor_view_idx(tensor_layout)]);
+#pragma unroll
+    for(int i = 0; i < 4; i++)
+    {
+        int64_t input_h = bound(in_y - 1 + i, Hin);
+#pragma unroll
+        for(int j = 0; j < 4; j++)
+        {
+            int64_t input_w = bound(in_x - 1 + j, Win);
+            tensor_layout_t<4> in_grad_layout(n, c, input_h, input_w);
+
+            atomicAdd(workspace + input_grad_tv.get_tensor_view_idx(in_grad_layout),
```

Comment:
As far as I understand your implementation, you used `atomicAdd` from HIP and it cannot support float16 and bfloat16, and after that you need to create another kernel `interpolateBicubicBackward_paste` to convert from float32 to (b)float16. That is not similar to modnn implementation, in modnn they implement a function called `atomic_add_g` which use `atomicCAS` from HIP. Take a look at this PR https://github.com/ROCm/MIOpen/pull/2883/ , seungmanhan has already implemented `hip_atomic.hpp` and then use `atomic_add_g` in `MIOpenGetitem.cpp`, which I think will produce better performance than this implementation. You don't need to fix it now, this is just a comment to notice that you may modify that when pushing to upstream repo  because in this branch there are no `hip_atomic.hpp` file.

Reply:
already knew about that atomic ops when https://github.com/ROCm/MIOpen/pull/2883 was not merged yet, I also tested in float32 case when use 2 kernel (bicubic + paste) and when use 1 kernel result in the same performance prove that paste kernel dont take much time. Although, I will add hic_atomic.hpp to avoid using workspace